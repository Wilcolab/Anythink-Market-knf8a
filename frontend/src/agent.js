import superagentPromise from "superagent-promise";
import _superagent from "superagent";

const superagent = superagentPromise(_superagent, global.Promise);

const BACKEND_URL = process.env.REACT_APP_BACKEND_URL;

const API_ROOT = `${BACKEND_URL}/api`;

const encode = encodeURIComponent;
const responseBody = (res) => res.body;

let token = null;
const tokenPlugin = (req) => {
  if (token) {
    req.set("authorization", `Token ${token}`);
  }
};

const requests = {
  del: (url) =>
    superagent.del(`${API_ROOT}${url}`).use(tokenPlugin).then(responseBody),
  get: (url) =>
    superagent.get(`${API_ROOT}${url}`).use(tokenPlugin).then(responseBody),
  put: (url, body) =>
    superagent
      .put(`${API_ROOT}${url}`, body)
      .use(tokenPlugin)
      .then(responseBody),
  post: (url, body) =>
    superagent
      .post(`${API_ROOT}${url}`, body)
      .use(tokenPlugin)
      .then(responseBody),
};

const Auth = {
  current: () => requests.get("/user"),
  login: (email, password) =>
    requests.post("/users/login", { user: { email, password } }),
  register: (username, email, password) =>
    requests.post("/users", { user: { username, email, password } }),
  save: (user) => requests.put("/user", { user }),
};

const Tags = {
  getAll: () => requests.get("/tags"),
};

const limit = (count, p) => `limit=${count}&offset=${p ? p * count : 0}`;
const omitSlug = (item) => Object.assign({}, item, { slug: undefined });
const Items = {
  all: (page) => requests.get(`/items?${limit(1000, page)}`),
  bySeller: (seller, page) =>
    requests.get(`/items?seller=${encode(seller)}&${limit(500, page)}`),
  byTag: (tag, page) =>
    requests.get(`/items?tag=${encode(tag)}&${limit(1000, page)}`),
  del: (slug) => requests.del(`/items/${slug}`),
  favorite: (slug) => requests.post(`/items/${slug}/favorite`),
  favoritedBy: (seller, page) =>
    requests.get(`/items?favorited=${encode(seller)}&${limit(500, page)}`),
  feed: () => requests.get("/items/feed?limit=10&offset=0"),
  get: (slug) => requests.get(`/items/${slug}`),
  unfavorite: (slug) => requests.del(`/items/${slug}/favorite`),
  update: (item) =>
    requests.put(`/items/${item.slug}`, { item: omitSlug(item) }),
  create: (item) => requests.post("/items", { item }),
};

const Comments = {
  create: (slug, comment) =>
    requests.post(`/items/${slug}/comments`, { comment }),
  delete: (slug, commentId) =>
    requests.del(`/items/${slug}/comments/${commentId}`),
  forItem: (slug) => requests.get(`/items/${slug}/comments`),
};

const Profile = {
  follow: (username) => requests.post(`/profiles/${username}/follow`),
  get: (username) => requests.get(`/profiles/${username}`),
  unfollow: (username) => requests.del(`/profiles/${username}/follow`),
};

const agentObj = {
  Items,
  Auth,
  Comments,
  Profile,
  Tags,
  setToken: (_token) => {
    token = _token;
  },
};

export default agentObj;
