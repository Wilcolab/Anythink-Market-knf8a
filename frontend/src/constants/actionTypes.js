export const APP_LOAD = "APP_LOAD";
export const REDIRECT = "REDIRECT";
export const ITEM_SUBMITTED = "ITEM_SUBMITTED";
export const SETTINGS_SAVED = "SETTINGS_SAVED";
export const DELETE_ITEM = "DELETE_ITEM";
export const SETTINGS_PAGE_UNLOADED = "SETTINGS_PAGE_UNLOADED";
export const HOME_PAGE_LOADED = "HOME_PAGE_LOADED";
export const HOME_PAGE_UNLOADED = "HOME_PAGE_UNLOADED";
export const ITEM_PAGE_LOADED = "ITEM_PAGE_LOADED";
export const ITEM_PAGE_UNLOADED = "ITEM_PAGE_UNLOADED";
export const ADD_COMMENT = "ADD_COMMENT";
export const DELETE_COMMENT = "DELETE_COMMENT";
export const ITEM_FAVORITED = "ITEM_FAVORITED";
export const ITEM_UNFAVORITED = "ITEM_UNFAVORITED";
export const SET_PAGE = "SET_PAGE";
export const APPLY_TAG_FILTER = "APPLY_TAG_FILTER";
export const CHANGE_TAB = "CHANGE_TAB";
export const PROFILE_PAGE_LOADED = "PROFILE_PAGE_LOADED";
export const PROFILE_PAGE_UNLOADED = "PROFILE_PAGE_UNLOADED";
export const LOGIN = "LOGIN";
export const LOGOUT = "LOGOUT";
export const REGISTER = "REGISTER";
export const LOGIN_PAGE_UNLOADED = "LOGIN_PAGE_UNLOADED";
export const REGISTER_PAGE_UNLOADED = "REGISTER_PAGE_UNLOADED";
export const ASYNC_START = "ASYNC_START";
export const ASYNC_END = "ASYNC_END";
export const EDITOR_PAGE_LOADED = "EDITOR_PAGE_LOADED";
export const EDITOR_PAGE_UNLOADED = "EDITOR_PAGE_UNLOADED";
export const ADD_TAG = "ADD_TAG";
export const REMOVE_TAG = "REMOVE_TAG";
export const UPDATE_FIELD_AUTH = "UPDATE_FIELD_AUTH";
export const UPDATE_FIELD_EDITOR = "UPDATE_FIELD_EDITOR";
export const FOLLOW_USER = "FOLLOW_USER";
export const UNFOLLOW_USER = "UNFOLLOW_USER";
export const PROFILE_FAVORITES_PAGE_UNLOADED =
  "PROFILE_FAVORITES_PAGE_UNLOADED";
export const PROFILE_FAVORITES_PAGE_LOADED = "PROFILE_FAVORITES_PAGE_LOADED";
