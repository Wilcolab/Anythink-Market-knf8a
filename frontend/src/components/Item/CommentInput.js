import React from "react";
import agent from "../../agent";
import { connect } from "react-redux";
import { ADD_COMMENT } from "../../constants/actionTypes";

const mapDispatchToProps = (dispatch) => ({
  onSubmit: (payload) => dispatch({ type: ADD_COMMENT, payload }),
});

class CommentInput extends React.Component {
  constructor() {
    super();
    this.state = {
      body: "",
    };

    this.setBody = (ev) => {
      this.setState({ body: ev.target.value });
    };

    this.createComment = async (ev) => {
      ev.preventDefault();
      agent.Comments.create(this.props.slug, {
        body: this.state.body,
      }).then((payload) => {
        this.props.onSubmit(payload);
      });
      this.setState({ body: "" });
    };
  }

  render() {
    return (
      <form className="card comment-form m-2" onSubmit={this.createComment}>
        <div className="card-block">
          <textarea
            className="form-control"
            placeholder="Write a comment..."
            value={this.state.body}
            onChange={this.setBody}
            rows="3"
          ></textarea>
        </div>
        <div className="card-footer">
          <img
            src={this.props.currentUser.image}
            className="user-pic mr-2"
            alt={this.props.currentUser.username}
          />
          <button className="btn btn-sm btn-primary" type="submit">
            Post Comment
          </button>
        </div>
      </form>
    );
  }
}

export default connect(() => ({}), mapDispatchToProps)(CommentInput);
