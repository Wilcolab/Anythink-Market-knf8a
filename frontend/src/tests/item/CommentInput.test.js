import { create } from "react-test-renderer";
import { mount } from "enzyme";
import configureMockStore from "redux-mock-store";
import CommentInput from "../../components/Item/CommentInput";
import agent from "../../agent";
import { ADD_COMMENT } from "../../constants/actionTypes";

const mockStore = configureMockStore();
agent.Comments.create = jest.fn();

describe("CommentInput component", () => {
  let store;

  beforeEach(() => {
    store = mockStore({});
  });

  it("Snapshot testing with no user", () => {
    const component = create(
      <CommentInput
        store={store}
        slug={"slug"}
        currentUser={{ username: "name", image: "" }}
      />
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("Submit text", () => {
    const user = { username: "name", image: "" };
    const component = mount(
      <CommentInput store={store} slug={"slug"} currentUser={user} />
    );
    const comment = { id: 1 };
    agent.Comments.create.mockResolvedValue(comment);

    const event = { target: { value: "sometext" } };
    component.find("textarea").simulate("change", event);
    component.find("form").simulate("submit");

    setImmediate(async () => {
      expect(store.getActions()).toHaveLength(1);
      expect(store.getActions()[0].type).toEqual(ADD_COMMENT);
      expect(await store.getActions()[0].payload).toEqual(comment);
    });
  });

  it("Clear text after submit", async () => {
    const user = { username: "name", image: "" };

    const component = mount(
      <CommentInput store={store} slug={"slug"} currentUser={user} />
    );

    const comment = { id: 1 };
    agent.Comments.create.mockResolvedValue(comment);

    const event = { target: { value: "sometext" } };
    component.find("textarea").simulate("change", event);
    component.find("form").simulate("submit");
    expect(component.find("textarea").text()).toHaveLength(0);
  });
});
