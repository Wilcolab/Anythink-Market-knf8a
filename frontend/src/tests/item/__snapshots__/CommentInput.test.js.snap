// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CommentInput component Snapshot testing with no user 1`] = `
<form
  className="card comment-form m-2"
  onSubmit={[Function]}
>
  <div
    className="card-block"
  >
    <textarea
      className="form-control"
      onChange={[Function]}
      placeholder="Write a comment..."
      rows="3"
      value=""
    />
  </div>
  <div
    className="card-footer"
  >
    <img
      alt="name"
      className="user-pic mr-2"
      src=""
    />
    <button
      className="btn btn-sm btn-primary"
      type="submit"
    >
      Post Comment
    </button>
  </div>
</form>
`;
